caproto/trio/server.py

```diff
@@ -111,27 +116,47 @@ def __init__(self, host, port, pvdb):
         task_status.started()
         await super().broadcast_beacon_loop()
 
-    async def server_accept_loop(self, addr, port, *, task_status):
-        with trio.socket.socket() as listen_sock:
-            self.log.debug('Listening on %s:%d', addr, port)
-            await listen_sock.bind((addr, port))
-            listen_sock.listen()
-
-            task_status.started()
-
-            while True:
-                client_sock, addr = await listen_sock.accept()
-                self.nursery.start_soon(self.tcp_handler, client_sock, addr)
+    async def server_accept_loop(self, listen_sock, *, task_status):
+            try:
+                listen_sock.listen()
+                task_status.started()
+                while True:
+                    client_sock, addr = await listen_sock.accept()
+                    self.nursery.start_soon(self.tcp_handler, client_sock, addr)
+            finally:
+                listen_sock.close()
 
     async def run(self, *, log_pv_names=False):
         'Start the server'
         self.log.info('Server starting up...')
         try:
             async with trio.open_nursery() as self.nursery:
-                for addr, port in ca.get_server_address_list(self.port):
-                    self.log.debug('Listening on %s:%d', addr, port)
+                # TODO Should this interface be configurable?
+                await self.beacon_sock.bind(('0.0.0.0', 0))
+                # Find a random port number that is free on all interfaces,
+                # and get a bound TCP socket with that port number on each
+                # interface.
+                tcp_sockets = {}  # maps interface to bound socket
+                stashed_ex = None
+                for port in ca.random_ports(100):
```

Comment:
Is it possible to push this up in server common?

Reply:
I started that way and pulled it apart when I realize that trio treats binding as asynchronous whereas curio and asyncio treat it as blocking. But, reflecting on your comment, I realized there are other places where only two of the three libs use code from common (and one is special-cased). So I moved the code in the base class for curio and asyncio and special-cased trio.